### PR TITLE
Add new tool: patch release cherry pick notification

### PR DESCRIPTION
**Pull request by cpanato**
#### What type of PR is this?

/kind feature

#### What this PR does / why we need it:

This is a tool that will run in a prow job and uses the https://raw.githubusercontent.com/kubernetes/website/main/data/releases/schedule.yaml published with the patch releases and will notify when the cherry pick deadline is close by sending an email to the k-dev mailing list.

Opened as WIP to collect feedback and see what we can improve

After approval and merge, will add the prow job to run the tool

/assign @saschagrunert @puerco @xmudrii @Verolop @justaugustus 
cc @kubernetes/release-engineering 

#### Which issue(s) this PR fixes:

Fixes partially https://github.com/kubernetes/release/issues/2174

#### Does this PR introduce a user-facing change?

```release-note
Add new tool: patch release cherry pick notification
```

### CODE REVIEW

File: cmd/patch-release-notify/cmd/root.go [hunk 1]
```diff
@@ -0,0 +1,294 @@
...
+	name           string
+	email          string
+	nomock         bool
+	logLevel       string
+}
+
+var opts = &options{}
+
+const (
+	sendgridAPIKeyEnvKey = "SENDGRID_API_KEY" // nolint: gosec
```
Comment (saschagrunert) [inline cmd/patch-release-notify/cmd/root.go:72]: Are we able to use this one in prow already?

**Review by saschagrunert — COMMENTED** [review]
I really like the idea of the early notification :+1:

File: cmd/patch-release-notify/cmd/root.go (hunk 1 above)
Comment (cpanato) [reply cmd/patch-release-notify/cmd/root.go:72]: we will need to add the SendGrid token in the prow.
and we will need to create a sendgrid token with a special account or use one of our personal accounts. i prefer to create a non-personal maybe using the release-managers, but don't know if that is possible in sendgrid

Comment (cpanato) [reply cmd/patch-release-notify/cmd/root.go:72]: i will finish this next week

**Comment by justaugustus** [comment]
Feel free to reassign me when this is ready!
/unassign

**Comment by cpanato** [comment]
sample email sent

Screenshot 2022-09-02 at 13 43 20

**Comment by cpanato** [comment]
will setup the job in a mock stage for now

cc @saschagrunert

File: cmd/schedule-builder/model/model.go [hunk 2]
```diff
@@ -32,6 +32,7 @@ type PreviousPatches struct {
 // Schedule struct to define the release schedule for a specific version
 type Schedule struct {
 	Release            string            `yaml:"release"`
+	ReleaseDate        string            `yaml:"releaseDate"`
 	Next               string            `yaml:"next"`
```
Comment (palnabarun) [inline cmd/schedule-builder/model/model.go:36]: https://github.com/kubernetes/website/pull/36609 adds more context to the `next` to make it homogenous with PreviousPatches structure.

File: cmd/patch-release-notify/cmd/root.go [hunk 3]
```diff
@@ -0,0 +1,294 @@
...
+		return fmt.Errorf("parsing template: %w", err)
+	}
+
+	var tmplBytes bytes.Buffer
+	err = tmpl.Execute(&tmplBytes, output)
+	if err != nil {
+		return fmt.Errorf("parsing values to the template: %w", err)
+	}
+
+	if shouldSendEmail {
```
Comment (palnabarun) [inline cmd/patch-release-notify/cmd/root.go:217]: Inverting the logic would improve the readability of the block.

```
if !shouldSendEmail {
  logrus.Info("No email is needed to send")
}

// email sending logic
```

File: cmd/patch-release-notify/cmd/root.go [hunk 4]
```diff
@@ -0,0 +1,294 @@
...
+	name           string
+	email          string
+	nomock         bool
+	logLevel       string
+}
+
+var opts = &options{}
+
+const (
+	sendgridAPIKeyEnvKey = "SENDGRID_API_KEY" // nolint: gosec
```
Comment (palnabarun) [inline cmd/patch-release-notify/cmd/root.go:72]: linter is failing here.

```
 cmd/patch-release-notify/cmd/root.go:72:44: whyNoLint: include an explanation for nolint directive (gocritic)
	sendgridAPIKeyEnvKey = "SENDGRID_API_KEY" // nolint: gosec 
```

**Comment by cpanato** [comment]
job: https://github.com/kubernetes/test-infra/pull/27489

**Comment by cpanato** [comment]
for now it will send the email to the test announcement mailing list to validate everything

**Comment by cpanato** [comment]
when you all have time can please take a look at this? pretty please 😇 

cc @saschagrunert @xmudrii @Verolop @palnabarun

File: cmd/patch-release-notify/README.md [hunk 5]
```diff
@@ -0,0 +1,16 @@
...
+
+This simple tool has the objective to send an notification email when we are closer to
+the patch release cycle to let people know that the cherry pick deadline is approaching.
+
+## Install
+
+The simplest way to install the `patch-release-notify` CLI is via `go get`:
+
+```
+$ go get k8s.io/release/cmd/patch-release-notify
```
Comment (saschagrunert) [inline cmd/patch-release-notify/README.md:11]:
```suggestion
$ go install k8s.io/release/cmd/patch-release-notify@latest
```

I think `go get` is deprecated for installing binaries.

File: cmd/patch-release-notify/cmd/root.go [hunk 6]
```diff
@@ -0,0 +1,295 @@
...
+	layout               = "2006-01-02"
+
+	schedulePathFlag = "schedule-path"
+	nameFlag         = "name"
+	emailFlag        = "email"
+	dayToalertFlag   = "days-to-alert"
+)
+
+var requiredFlags = []string{
+	schedulePathFlag,
```
Comment (saschagrunert) [inline cmd/patch-release-notify/cmd/root.go:82]: Could we point that to a default remote location without requiring the path to be locally available?

**Comment by justaugustus** [comment]
@cpanato – should we resurrect this PR?

**Comment by cpanato** [comment]
@justaugustus yes!! i know we are discussing which email service to use, so I would like to align that. do you have any updates @jimangel @ameukam? what is the email service?

then I can update this PR and we can move forward

**Comment by jimangel** [comment]
> what is the email service?

The solution I proposed is for manual sending (releng folks cutting releases). This seems like it's attempting to be a more automated solution? If so, we'd need to evaluate a different bulk-sending service like @ameukam mentioned.

TL;DR: I don't think there's a solution yet for automated emails.

We also might be able to modify my solution to use an offline / refresh token, but then we'd have to figure out which account we'd want to authorize to allow "sending email on behalf."

**Comment by cpanato** [comment]
/remove-lifecycle rotten

working on that

**Comment by cpanato** [comment]
closing in favor of https://github.com/kubernetes/release/pull/3627
